test/Microsoft.AspNetCore.SignalR.Tests/TestClient.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.IO;
+using System.IO.Pipelines;
+using System.Security.Claims;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Channels;
+using Microsoft.AspNetCore.Sockets;
+using Microsoft.AspNetCore.Sockets.Internal;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.SignalR.Tests
+{
+    public class TestClient : IDisposable
+    {
+        private static int _id;
+        private IInvocationAdapter _adapter;
+        private CancellationTokenSource _cts;
+
+        public Connection Connection;
+        public IChannelConnection<Message> Application { get; }
+        public Task Connected => Connection.Metadata.Get<TaskCompletionSource<bool>>("ConnectedTask").Task;
+
+        public TestClient(IServiceProvider serviceProvider, string format = "json")
+        {
+            var transportToApplication = Channel.CreateUnbounded<Message>();
+            var applicationToTransport = Channel.CreateUnbounded<Message>();
+
+            Application = ChannelConnection.Create(input: applicationToTransport, output: transportToApplication);
+            var transport = ChannelConnection.Create(input: transportToApplication, output: applicationToTransport);
+
+            Connection = new Connection(Guid.NewGuid().ToString(), transport);
+            Connection.Metadata["formatType"] = format;
+            Connection.User = new ClaimsPrincipal(new ClaimsIdentity(new[] { new Claim(ClaimTypes.Name, Interlocked.Increment(ref _id).ToString()) }));
+            Connection.Metadata["ConnectedTask"] = new TaskCompletionSource<bool>();
+
+            var invocationAdapter = serviceProvider.GetService<InvocationAdapterRegistry>();
+            _adapter = invocationAdapter.GetInvocationAdapter(format);
+
+            _cts = new CancellationTokenSource();
+        }
+
+        public async Task<T> Invoke<T>(string methodName, params object[] args)
+        {
+            await Invoke(methodName, args);
+
+            return await Read<T>();
+        }
+
+        public async Task Invoke(string methodName, params object[] args)
+        {
+            var stream = new MemoryStream();
+            await _adapter.WriteMessageAsync(new InvocationDescriptor
+            {
+                Arguments = args,
+                Method = methodName
+            },
+            stream);
+
+            var buffer = ReadableBuffer.Create(stream.ToArray()).Preserve();
+            await Application.Output.WriteAsync(new Message(buffer, Format.Binary, endOfMessage: true));
+        }
+
+        public async Task<T> Read<T>()
```

Comment:
e.g. Client1 broadcasts to all clients, how else do we let Client2 see the message?